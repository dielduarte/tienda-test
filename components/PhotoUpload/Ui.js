import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export const Container = styled.div`
  padding: 30px 20px 10px 20px;
`;

export const Label = styled.label`
  font-size: 14px;
`;

export const DashedContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: dashed 2px ${props => props.theme.gray};
  width: calc(25% - 20px);
  height: 150px;
  margin: 20px 20px 0 0;

  input {
    visibility: hidden;
    width: 0;
  }

  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }

  &:hover {
    cursor: ${props => (props.withHover ? 'pointer' : 'default')};
  }
`;

export const Icon = styled(FontAwesomeIcon)`
  font-size: 40px;
  color: ${props => props.theme.gray};
`;
