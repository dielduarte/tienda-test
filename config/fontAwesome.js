import { library } from '@fortawesome/fontawesome-svg-core';
import {
  faExternalLinkAlt,
  faEllipsisV,
  faBars,
  faAngleRight,
  faCamera
} from '@fortawesome/free-solid-svg-icons';

library.add(faExternalLinkAlt, faEllipsisV, faBars, faAngleRight, faCamera);
