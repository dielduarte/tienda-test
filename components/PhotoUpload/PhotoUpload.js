import React from 'react';
import { withHandlers, withState, lifecycle, compose } from 'recompose';
import { func, array, oneOf, string, node } from 'prop-types';

import { Row } from '../Flex';
import { Container, Label, DashedContainer, Icon } from './Ui';

export const openFileSearch = props => () => {
  props.uploadInputRef.current.click();
};

export const onUploadFinished = (props, image) => {
  const { setPhotos, photos, onUploadSuccess } = props;
  const newPhotos = [...photos, image];

  setPhotos(newPhotos);
  onUploadSuccess(newPhotos);
};

export const uploadPhoto = (Reader, cb) => props => () => {
  const file = props.uploadInputRef.current.files[0];

  Reader.addEventListener('load', () => cb(props, Reader.result));

  if (file) {
    Reader.readAsDataURL(file);
  }
};

export function renderPhotos(photos) {
  return photos.map((photo, key) => (
    <DashedContainer key={key}>
      <img src={photo} />
    </DashedContainer>
  ));
}

function PhotoUpload({
  label,
  openFileSearch,
  uploadInputRef,
  uploadPhoto,
  photos
}) {
  return (
    <Container>
      <Label>{label}</Label>
      <Row wrap="true">
        {renderPhotos(photos)}
        <DashedContainer onClick={openFileSearch} withHover>
          <input type="file" ref={uploadInputRef} onChange={uploadPhoto} />
          <Icon icon="camera" />
        </DashedContainer>
      </Row>
    </Container>
  );
}

PhotoUpload.propsType = {
  label: string.isRequired,
  openFileSearch: func.isRequired,
  uploadInputRef: oneOf([string, node]).isRequired,
  uploadPhoto: func.isRequired,
  photos: array.isRequired,
  onUploadSuccess: func.isRequired
};

export default compose(
  withState('uploadInputRef', 'setUploadInputRef', null),
  withState('photos', 'setPhotos', []),
  lifecycle({
    componentDidMount() {
      this.props.setUploadInputRef(React.createRef());
    }
  }),
  withHandlers({
    openFileSearch,
    uploadPhoto: uploadPhoto(new FileReader(), onUploadFinished)
  })
)(PhotoUpload);
