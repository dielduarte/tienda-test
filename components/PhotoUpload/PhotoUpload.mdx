---
name: PhotoUpload
---

import '../../config/fontAwesome';
import { Playground, PropsTable } from 'docz'
import PhotoUpload from './PhotoUpload';

# PhotoUpload

<Playground>
  <PhotoUpload onUploadSuccess={(photos) => console.log(photos)}/>
</Playground>