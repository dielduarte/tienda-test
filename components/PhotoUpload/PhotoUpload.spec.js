import { shallow, mount } from 'enzyme';
import 'jest-styled-components';

import '../../config/fontAwesome';

import PhotoUpload, {
  openFileSearch,
  onUploadFinished,
  uploadPhoto,
  renderPhotos
} from './PhotoUpload';

describe('PhotoUpload', () => {
  it('should render correctly', () => {
    const wrapper = mount(
      <PhotoUpload
        label="label teste"
        onUploadSuccess={photos => console.log(photos)}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the click event of the ref element', () => {
    const props = {
      uploadInputRef: {
        current: { click: jest.fn() }
      }
    };

    openFileSearch(props)();
    expect(props.uploadInputRef.current.click).toHaveBeenCalled();
  });

  it('should call setPhotos and onUploadSuccess props passing the new photos array', () => {
    const props = {
      setPhotos: jest.fn(),
      photos: ['cdn.nuvem.com/image1.example.jpg'],
      onUploadSuccess: jest.fn()
    };

    const newImage = 'cdn.nuvem.com/image2.example.jpg';

    const newPhotosArray = [...props.photos, newImage];

    onUploadFinished(props, newImage);
    expect(props.setPhotos).toHaveBeenCalledWith(newPhotosArray);
    expect(props.onUploadSuccess).toHaveBeenCalledWith(newPhotosArray);
  });

  it('should render the photo list', () => {
    const photoList = [
      'https://cdn.nuvem.com/image1.png',
      'https://cdn.nuvem.com/image2.png'
    ];

    let wrapper = shallow(
      <PhotoUpload
        label="label teste"
        onUploadSuccess={photos => console.log(photos)}
      />
    ).dive();

    wrapper.props().setPhotos(photoList);
    expect(wrapper).toMatchSnapshot();

    wrapper = renderPhotos(photoList);
    expect(wrapper).toMatchSnapshot();
  });

  it('should upload the photo and call the callback function passing the image', () => {
    const file = 'fakefile.png';

    const props = {
      uploadInputRef: {
        current: {
          files: [file]
        }
      }
    };

    const onUploadFinished = jest.fn();
    const eventMap = {};

    const FileReader = {
      readAsDataURL: jest.fn(() => eventMap['load']()),
      addEventListener: (type, cb) => {
        eventMap[type] = cb;
      },
      result: file
    };

    uploadPhoto(FileReader, onUploadFinished)(props)();

    expect(FileReader.readAsDataURL).toHaveBeenCalledWith(file);
    expect(onUploadFinished).toHaveBeenCalledWith(props, file);
  });

  it(`should not call the callback function if haven't image`, () => {
    const file = null;

    const props = {
      uploadInputRef: {
        current: {
          files: [file]
        }
      }
    };

    const onUploadFinished = jest.fn();
    const eventMap = {};

    const FileReader = {
      readAsDataURL: jest.fn(() => eventMap['load']()),
      addEventListener: (type, cb) => {
        eventMap[type] = cb;
      },
      result: file
    };

    uploadPhoto(FileReader, onUploadFinished)(props)();

    expect(FileReader.readAsDataURL).not.toHaveBeenCalledWith(file);
    expect(onUploadFinished).not.toHaveBeenCalledWith(props, file);
  });
});
