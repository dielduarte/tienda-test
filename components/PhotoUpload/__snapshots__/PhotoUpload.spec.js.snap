// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoUpload should render correctly 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 {
  padding: 30px 20px 10px 20px;
}

.c1 {
  font-size: 14px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: dashed 2px;
  width: calc(25% - 20px);
  height: 150px;
  margin: 20px 20px 0 0;
}

.c3 input {
  visibility: hidden;
  width: 0;
}

.c3 img {
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.c3:hover {
  cursor: pointer;
}

.c4 {
  font-size: 40px;
}

@media (max-width:480px) {
  .c3 {
    width: calc(50% - 20px);
  }
}

<withState(withState(lifecycle(withHandlers(PhotoUpload))))
  label="label teste"
  onUploadSuccess={[Function]}
>
  <withState(lifecycle(withHandlers(PhotoUpload)))
    label="label teste"
    onUploadSuccess={[Function]}
    setUploadInputRef={[Function]}
    uploadInputRef={
      Object {
        "current": <input
          type="file"
        />,
      }
    }
  >
    <lifecycle(withHandlers(PhotoUpload))
      label="label teste"
      onUploadSuccess={[Function]}
      photos={Array []}
      setPhotos={[Function]}
      setUploadInputRef={[Function]}
      uploadInputRef={
        Object {
          "current": <input
            type="file"
          />,
        }
      }
    >
      <withHandlers(PhotoUpload)
        label="label teste"
        onUploadSuccess={[Function]}
        photos={Array []}
        setPhotos={[Function]}
        setUploadInputRef={[Function]}
        uploadInputRef={
          Object {
            "current": <input
              type="file"
            />,
          }
        }
      >
        <PhotoUpload
          label="label teste"
          onUploadSuccess={[Function]}
          openFileSearch={[Function]}
          photos={Array []}
          setPhotos={[Function]}
          setUploadInputRef={[Function]}
          uploadInputRef={
            Object {
              "current": <input
                type="file"
              />,
            }
          }
          uploadPhoto={[Function]}
        >
          <styled.div>
            <div
              className="c0"
            >
              <styled.label>
                <label
                  className="c1"
                >
                  label teste
                </label>
              </styled.label>
              <styled.div
                wrap="true"
              >
                <div
                  className="c2"
                  wrap="true"
                >
                  <styled.div
                    onClick={[Function]}
                    withHover={true}
                  >
                    <div
                      className="c3"
                      onClick={[Function]}
                    >
                      <input
                        onChange={[Function]}
                        type="file"
                      />
                      <Styled(FontAwesomeIcon)
                        icon="camera"
                      >
                        <FontAwesomeIcon
                          border={false}
                          className="c4"
                          fixedWidth={false}
                          flip={null}
                          icon="camera"
                          inverse={false}
                          listItem={false}
                          mask={null}
                          pull={null}
                          pulse={false}
                          rotation={null}
                          size={null}
                          spin={false}
                          symbol={false}
                          title=""
                          transform={null}
                        >
                          <svg
                            aria-hidden="true"
                            className="svg-inline--fa fa-camera fa-w-16 c4"
                            data-icon="camera"
                            data-prefix="fas"
                            role="img"
                            style={Object {}}
                            viewBox="0 0 512 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M512 144v288c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V144c0-26.5 21.5-48 48-48h88l12.3-32.9c7-18.7 24.9-31.1 44.9-31.1h125.5c20 0 37.9 12.4 44.9 31.1L376 96h88c26.5 0 48 21.5 48 48zM376 288c0-66.2-53.8-120-120-120s-120 53.8-120 120 53.8 120 120 120 120-53.8 120-120zm-32 0c0 48.5-39.5 88-88 88s-88-39.5-88-88 39.5-88 88-88 88 39.5 88 88z"
                              fill="currentColor"
                              style={Object {}}
                            />
                          </svg>
                        </FontAwesomeIcon>
                      </Styled(FontAwesomeIcon)>
                    </div>
                  </styled.div>
                </div>
              </styled.div>
            </div>
          </styled.div>
        </PhotoUpload>
      </withHandlers(PhotoUpload)>
    </lifecycle(withHandlers(PhotoUpload))>
  </withState(lifecycle(withHandlers(PhotoUpload)))>
</withState(withState(lifecycle(withHandlers(PhotoUpload))))>
`;

exports[`PhotoUpload should render the photo list 1`] = `
<lifecycle(withHandlers(PhotoUpload))
  label="label teste"
  onUploadSuccess={[Function]}
  photos={
    Array [
      "https://cdn.nuvem.com/image1.png",
      "https://cdn.nuvem.com/image2.png",
    ]
  }
  setPhotos={[Function]}
  setUploadInputRef={[Function]}
  uploadInputRef={null}
/>
`;

exports[`PhotoUpload should render the photo list 2`] = `
Array [
  <styled.div>
    <img
      src="https://cdn.nuvem.com/image1.png"
    />
  </styled.div>,
  <styled.div>
    <img
      src="https://cdn.nuvem.com/image2.png"
    />
  </styled.div>,
]
`;
